### ISSUE TITLE
Regalloc conflict produces internal error when function is inlined

### ISSUE THREAD

**Issue by sarranz** [jasmin-lang/jasmin#483]
The following gives `internal compilation error in function main: Variable allocation: i (defined at "ble.jazz", line 16 (12-13)) and i (defined at "ble.jazz", line 16 (12-13)) : variable already set`
```
fn f(reg ptr u8[1] p, reg u64 i) -> reg ptr u8[1] {
    while (i < 1) {
        p[(int) i] = 0;
        i += 1;
    }
    return p;
}

export
fn main () -> reg u8 {
    stack u8[1] c;
    reg ptr u8[1] p;
    p = c;

    reg u64 i;
    i = 0;

    while (i < 1) {
        #inline
        p = f(p, i);
        i += 1;
    }

    reg u8 tmp;
    tmp = p[0];
    return tmp;
}
```
I believe it's because `f` increases `i` but `main` does not see this. If I remove the `#inline`, then I get a register allocation error saying that these two `i`s conflict, this is what I'd expect. Naturally, it works if I introduce `i = i` at the beginning of `f`, or return `i` in `f`.
Is this really an internal error, then?

**Comment by vbgl** [comment]
Well spotted. Here is a smaller example:

~~~
inline fn f(reg u64 j) {
    while (j < 1) {
        j += 1;
    }
}

export
fn main (reg u64 i) {
    while (i < 1) {
        f(i);
        i += 1;
    }
}
~~~

I think the issue is in the “renaming” step that blindly trusts the “rename” tag attached to assignments that are introduced by the “inlining” phase.

A simple workaround is to add an explicit assignment “i = i;” at the beginning of the inlined function.

**Comment by eponier** [comment]
With @clebreto, we encountered the same error while writing code for Dilithium.
```
nline fn __inline_f(reg u32 inline_in) -> reg u32 {
  reg u32 inline_out;
  if (inline_in > 7) {
    inline_out = inline_in + 3;
  }
  return inline_out;
}

export
fn add(reg u32 arg0, reg u32 arg1) -> reg u32 {

  reg u32 i;
  i = 0;
  while { i = i  + 1; } ( i < 2 ) {
    arg1 = __inline_f(arg0);
    if (arg0 > 3) {
        arg1 = __inline_f(arg0);
    }
  }

  reg u32 res;
  res = arg1;
  return res;
}
```
It seems a bit different, in the sense that without the `inline` keyword, it compiles fine (at least on arm-m4).

**Comment by eponier** [comment]
We minimized the original code, I'll have a look to check if the code was already strange or if we did the minimization wrongly.

**Comment by eponier** [comment]
Indeed, the potentially uninitialized variable was already there. Sorry for the noise (even if, of course, the current message could be improved).
